### DISCUSSION TITLE
Rate limiting clarification

### DISCUSSION THREAD

**Discussion by shaunkh** [taskforcesh/bullmq#1540] [Q&A] [answered]
The docs state that "The rate limiter is global, so if you have for example 10 workers for one queue with the above settings, still only 10 jobs will be processed by second."

Does this mean if you have 3 workers each individually on 3 different queues, you can set a limiter for each worker?
Or is it still global? i.e. I set a limiter on 1st worker to be 5 and 2nd worker to be 10. Are all workers limited to 5 now even though they are on different queues?

**Comment by manast** [comment, marked as answer]
"The rate limiter is global, so if you have for example 10 workers **for one queue** with the above settings, still only 10 jobs will be processed by second."

**Reply by shaunkh** [reply]
Sorry but you just reiterated what the docs said. Doesn't really help at all or answer the question because that line only talks about if you have one queue.

I just want to be 100% sure when using bullmq.

**Reply by manast** [reply]
You define the global rate limiter on one Queue so yes, it affects only that queue.

**Reply by shaunkh** [reply]
Ahh ok thank you

**Comment by dschnare** [comment]
Hey @manast I know you already answered this discussion, and I fully understand the rate limiter is global. However, what is the behaviour of the rate limiter in the scenario where a worker process is scaled horizontally with rate limiter settings of 10/sec, then an update to the worker code is made and a progressive deployment occurs OR a new worker process is brought online with rate limiter settings of 20/sec. All workers are processors of the same queue. Basically what happens when you have workers with different rate limiter settings that process the same queue?

**Reply by manast** [reply]
Mixing different rate-limit options for the same queue will sometimes rate limit at one of the worker options and sometimes the other, it all depends on which worker actually gets to limit the job, so it will be quite random. So for consistent behavior, try to always use the same settings for all workers running for a given queue.

**Reply by dschnare** [reply]
Got it. So for something like blue-green worker services, the time that the rate limiting options are different should be minimal, and you really do want to ensure you transition to green as soon as practical to avoid this random rate limiting behaviour. Understood. Thanks for the help.

**Reply by manast** [reply]
Yes, to clarify, it will be random between the different rate-limit settings, not a random rate-limiting. So if you have two limits 10/s and 30/s, sometimes it will rate at 10 other times at 30, but I think it will be heavily biased to 10 actually since it is the most restrictive limit.
